concordium-std-derive/src/lib.rs

```diff
@@ -1202,6 +1222,57 @@ fn schema_type_fields(fields: &syn::Fields) -> syn::Result<proc_macro2::TokenStr
     }
 }
 
+/// We reserve a number of error codes for custom errors, such as ParseError,
+/// that are provided by concordium-std. The error codes for user-defined error
+/// types will have this number as an offset.
+const RESERVED_ERROR_CODES: u32 = 100;
+
+/// Derive the conversion of enums that represent error types into the Reject
+/// struct which can be used as the error type of init and receive functions.
+/// Creating custom enums for error types can provide meaningful error messages
+/// to the user of the smart contract.
+///
+/// Please note:
+///   - At the moment, we can only derive fieldless enums.
+///   - The error-type enum needs to derive (or implement) the Copy and Clone
+///     traits.
+///
+/// ### Example
+/// ```ignore
+/// #[derive(Clone, Copy, Reject)]
+/// enum MyError {
+///     IllegalState,
+///     WrongSender,
+///     // TimeExpired(time: Timestamp), /* currently not supported */
+///     ...
+/// }
+/// ```ignore
+/// #[receive(contract = "my_contract", name = "some_receive")]
+/// fn receive<A: HasActions>(ctx: &impl HasReceiveContext, state: &mut MyState)
+/// -> Result<A, MyError> {...} ```
+#[proc_macro_derive(Reject)]
+pub fn reject_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).unwrap();
+    impl_reject_derive(&ast)
+}
+
+fn impl_reject_derive(ast: &syn::DeriveInput) -> TokenStream {
+    let enum_name = &ast.ident;
+    let exceeded_error_code_limit = format!(
+        "Error code should not exceed {} (i32::MAX minus reserved error-code offset).",
+        i32::MAX - RESERVED_ERROR_CODES as i32
+    );
+
+    let gen = quote! {
+        impl From<#enum_name> for Reject {
+            fn from(e: #enum_name) -> Self {
+                Reject { error_code: (e as u32).checked_add(#RESERVED_ERROR_CODES).expect(#exceeded_error_code_limit) }
```

Comment:
I think for fieldless enums this is a good (and cheap, which is important) option.

For enums with fields I think we'll want to generate an explicit match statement so we can handle sub-enums better as we discussed yesterday.

We need to do some experimentation to see what is most ergonomic, and least error prone for the user.